C_GetAttributeValue implementation and other improvements

EDVTAZ opened the pull request:
Signed-off-by: Gabor Szekely <szvgabor@gmail.com>

- Implemented C_GetAttributeValue command, but access control needs
to be implemented before this can be used

- Modified code related to C_FindObjectsInit and friends,
so it will match all objects if there are no attributes supplied
in the template, as specified by the standard.

- Added SKS_CKFT_LOGIN_REQUIRED to token flags as default,
because otherwise openssh wouldn't ask for pin number and fail,
since the private key is only available to authenticated users.
I'm not 100% sure that this is the right way to add this flag,
comments are welcome.

- Added automatic CKA_ID generation to the generate key pair
code, because otherwise openssh wouldn't use the generated
key pair. The standard mentions that public and private key
objects may be paired based on their id (although doesn't
require this to be enforced), I assume this is why openssh
only works with key pairs that have CKA_ID.

<!--
    If you are new to submitting pull requests to OP-TEE, then please have a
    look at the list below and tick them off before submitting the pull request.

    1. Read our contribution guidelines:
         documentation/github.md.

    2. Read the contribution section in Notice.md and pay extra attention to the
       "Developer Certificate of Origin" part:
         https://github.com/OP-TEE/optee_os/blob/master/Notice.md#contributions.

    3. You should run checkpatch preferably before submitting the pull request.

    4. When everything has been reviewed, you will need to squash, rebase and
       add tags like `Reviewed-by`, `Acked-by`, `Tested-by` etc.

    NOTE: This comment will not be shown in the pull request, so no harm keeping
    it, but feel free to remove it if you like.
-->

EDVTAZ committed:
C_GetAttributeValue implementation and other improvements

Signed-off-by: Gabor Szekely <szvgabor@gmail.com>

- Implemented C_GetAttributeValue command, but access control needs
to be implemented before this can be used

- Modified code related to C_FindObjectsInit and friends,
so it will match all objects if there are no attributes supplied
in the template, as specified by the standard.

- Added SKS_CKFT_LOGIN_REQUIRED to token flags as default,
because otherwise openssh wouldn't ask for pin number and fail,
since the private key is only available to authenticated users.
I'm not 100% sure that this is the right way to add this flag,
comments are welcome.

- Added automatic CKA_ID generation to the generate key pair
code, because otherwise openssh wouldn't use the generated
key pair. The standard mentions that public and private key
objects may be paired based on their id (although doesn't
require this to be enforced), I assume this is why openssh
only works with key pairs that have CKA_ID.

EDVTAZ commented:
Hi @etienne-lms ,
I have been testing the sks with pkcs11-tool and openssh, and made some modifications in it based on that. The details about the changes are in the commit message. Some of the new features I added are not yet complete, but I wanted to show you what I'm working on. Any comments are appreciated. I put a PR on the optee_client repository as well.
One more thing that is not yet in the code because I haven't come up with a solution yet: when openssh starts the authentication with the server it uses the `CKM_RSA_PKCS` mechanism with the `C_Sign` method. Currently this is translated to `TEE_ALG_RSAES_PKCS1_V1_5` on the TEE side, which fails, because this algorithm only supports encryption and decryption. The problem is, however that even though at first this seems easily fixable by translating above mentioned case into `TEE_ALG_RSASSA_PKCS1_V1_5`, it turns out there is no such algorithm in the TEE API, only ones with hash functions integrated, like `TEE_ALG_RSASSA_PKCS1_V1_5_SHA1`. Do you have any ideas on how to solve this problem? Could I maybe use the libtomcrypt library that OPTEE uses in the background?

EDVTAZ committed:
C_GetAttributeValue better access control and rv

Signed-off-by: Gabor Szekely <szvgabor@gmail.com>

EDVTAZ committed:
Added more access check in entry_get_attribute_value

Signed-off-by: Gabor Szekely <szvgabor@gmail.com>

EDVTAZ committed:
Init return value in sanitize_class_and_type

Signed-off-by: Gabor Szekely <szvgabor@gmail.com>

This is needed so it doesn't fail empty templates.

EDVTAZ committed:
Add TEE_ALG_RSASSA_PKCS1_V1_5 algorithm

Signed-off-by: Gabor Szekely <szvgabor@gmail.com>

Note that TEE_ALG_RSASSA_PKCS1_V1_5 is not part of the GPD TEE
standard. It was added, because I couldn't find a way to implement
the CKM_RSA_PKCS mechanism signing and verifying through the TEE
API. The TEE_ALG_RSASSA_PKCS1_V1_5 currently uses EMSA padding and
signs the message without hashing it.

EDVTAZ commented:
Right now, I can use ssh with sks backing the client (`ssh -I /usr/lib/libsks.so`). I added the `TEE_ALG_RSASSA_PKCS1_V1_5` to the TEE API, although I don't know if this can be merged. It is not part of the GPD TEE specification, but I don't think it breaks anything either. I also updated the `rsa_sign_hash.c` and `rsa_verify_hash.c` files, because the ones in OP-TEE were outdated, and there was no option to sign without hashing.

etienne-lms commented on core/lib/libtomcrypt/src/pk/rsa/rsa_sign_hash.c:
```diff
@@ -42,7 +42,6 @@
   @file rsa_sign_hash.c
   RSA PKCS #1 v1.5 and v2 PSS sign hash, Tom St Denis and Andreas Lange
 */
-
```
minor: can be discarded.

EDVTAZ replied:
I replaced this file with a newer version, that's the reason these comments are a bit different. I will go through it again and discard these changes.

etienne-lms commented on core/lib/libtomcrypt/src/pk/rsa/rsa_sign_hash.c:
```diff
@@ -75,15 +74,21 @@ int rsa_sign_hash_ex(const unsigned char *in,       unsigned long  inlen,
    }
 
    if (padding == LTC_PKCS_1_PSS) {
-     /* valid prng and hash ? */
```
minor: discard change

EDVTAZ replied:
Same as the other one, I will discard.

EDVTAZ replied:
Actually I'm not sure about this one. If I understand this correctly, this block only checks the prng, the next one checks hash for every padding (except for LTC_PKCS_1_V1_5_NA1, where there is no need for a hash). It seems to me that the older version of the comment is incorrect.

etienne-lms replied:
sorry. ok with your change.

etienne-lms commented on core/lib/libtomcrypt/src/pk/rsa/rsa_sign_hash.c:
```diff
@@ -75,15 +74,21 @@ int rsa_sign_hash_ex(const unsigned char *in,       unsigned long  inlen,
      }
+   }
+
+   if (padding != LTC_PKCS_1_V1_5_NA1) {
```
Previous sequence did not check hash for `LTC_PKCS_1_V1_5`. If this is an error, we should fix through a specific commit of propose straight to optee_os.

EDVTAZ replied:
I think this is correct the way it is, the hash for `LTC_PKCS_1_V1_5` is being checked in this block. Also, see my comment above.

etienne-lms replied:
my fault. you're right.

etienne-lms commented on core/lib/libtomcrypt/src/pk/rsa/rsa_verify_hash.c:
```diff
@@ -203,6 +218,6 @@ int rsa_verify_hash_ex(const unsigned char *sig,      unsigned long siglen,
 
 #endif /* LTC_MRSA */
 
-/* $Source: /cvs/libtom/libtomcrypt/src/pk/rsa/rsa_verify_hash.c,v $ */
```
why this change?

EDVTAZ replied:
I will discard.

etienne-lms commented on lib/libutee/include/tee_api_defines.h:
```diff
@@ -134,6 +134,7 @@
 #define TEE_ALG_DES3_CBC_NOPAD                  0x10000113
 #define TEE_ALG_DES3_CBC_MAC_NOPAD              0x30000113
 #define TEE_ALG_DES3_CBC_MAC_PKCS5              0x30000513
+#define TEE_ALG_RSASSA_PKCS1_V1_5               0x70007830
```
This would rather go in **tee_api_defines_extensions.h** as [hkdf extensions](https://github.com/OP-TEE/optee_os/blob/master/lib/libutee/include/tee_api_defines_extensions.h#L9).

EDVTAZ replied:
I wasn't aware of that, I will move it.

etienne-lms commented on ta_services/secure_key_services/src/processing_asymm.c:
```diff
@@ -138,6 +139,11 @@ static uint32_t sks2tee_algorithm(uint32_t *tee_id,
+	if (*tee_id == TEE_ALG_RSAES_PKCS1_V1_5 &&
+		(function == SKS_FUNCTION_SIGN ||
+		 function == SKS_FUNCTION_VERIFY))
+		*tee_id = TEE_ALG_RSASSA_PKCS1_V1_5;
```
I don't think it is nice to switch the algorithm id.
If `TEE_ALG_RSASSA_PKCS1_V1_5` is dedicated to sing/verify a pre-hashed chunk caller should have set `*tee_id` to the right ID.

EDVTAZ replied:
This is part of the function that translates the SKS_CKM_* to tee_id, and SKS_CKM_RSA_PKCS can be either TEE_ALG_RSAES_PKCS1_V1_5 or TEE_ALG_RSASSA_PKCS1_V1_5 based on whether it is called with C_Encrypt/C_Decrypt or C_Sign/C_Verify. I will place this chunk closer to the code that does the initial translation, that way it will be more intuitive.

etienne-lms replied:
ok, fine. I did not understand the sequence at the time.

etienne-lms commented on ta_services/secure_key_services/src/pkcs11_attributes.h:
```diff
@@ -11,6 +11,8 @@
 
 #include "serializer.h"
 
+#define CK_UNAVAILABLE_INFORMATION (~0UL)
```
I found that  `SKS_UNDEFINED_ID` duplicates `CK_UNAVAILABLE_INFORMATION` in the TA API :(

I propose to define  `SKS_UNDEFINED_ID` as an alias to a new `SKS_CK_UNAVAILABLE_INFORMATION` (all `CK_xxx` are prefixed `SKS_` in the TA API, more flexible to prepare integration of the TA in op-tee mainline).

EDVTAZ replied:
Ok, I think I can do that.

etienne-lms commented on ta_services/secure_key_services/src/object.c:
```diff
@@ -731,19 +758,53 @@ uint32_t entry_get_attribute_value(uintptr_t tee_session, TEE_Param *ctrl,
+		if (!attribute_is_exportable(cli_ref, obj)) {
+			cli_ref->size = CK_UNAVAILABLE_INFORMATION;
+			attr_sensitive = 1;
+		}
```
I think you should add a `continue;` in that block; don't try to get the data.

EDVTAZ replied:
Yes, I will do that.

etienne-lms commented on ta_services/secure_key_services/src/object.c:
```diff
@@ -731,19 +758,53 @@ uint32_t entry_get_attribute_value(uintptr_t tee_session, TEE_Param *ctrl,
+			attr_type_invalid = 1;
+		}
+		if (rv == SKS_SHORT_BUFFER)
+			buffer_too_small = 1;
```
If `rv` is not `NOT_FOUND`, `SHORT_BUFFER` or `OK`, function should fail straight.

EDVTAZ replied:
The pkcs11 standard asks that every attribute should be processed:

> Note that the error codes CKR_ATTRIBUTE_SENSITIVE, CKR_ATTRIBUTE_TYPE_INVALID, and CKR_BUFFER_TOO_SMALL do not denote true errors for C_GetAttributeValue. If a call to C_GetAttributeValue returns any of these three values, then the call MUST nonetheless have processed every attribute in the template supplied to C_GetAttributeValue. Each attribute in the template whose value can be returned by the call to C_GetAttributeValue will be returned by the call to C_GetAttributeValue.

etienne-lms commented on ta_services/secure_key_services/src/object.c:
```diff
@@ -725,6 +725,12 @@ uint32_t entry_get_attribute_value(uintptr_t tee_session, TEE_Param *ctrl,
 	if (!obj)
 		return SKS_BAD_PARAM;
```
Ouch! Looks like a goto bail is missing here. Can you fix while you're around?

etienne-lms commented on ta_services/secure_key_services/src/sks_helpers.c:
```diff
@@ -132,6 +132,24 @@ static const struct processing_id __maybe_unused processing_ids[] = {
+	SKS_PROCESSING_ID(SKS_CKM_RSA_PKCS_KEY_PAIR_GEN),
+	SKS_UNSUPPORTED_PROCESSING_ID(SKS_CKM_RSA_PKCS),
+	SKS_PROCESSING_ID(SKS_CKM_RSA_9796),
+	SKS_PROCESSING_ID(SKS_CKM_RSA_X_509),
```
this is supported?

EDVTAZ replied:
I just added these so I can find what mechanisms client applications needed. I will look through them more carefully and remove the ones that are incorrect.

etienne-lms commented on ta_services/secure_key_services/src/sks_helpers.c:
```diff
@@ -132,6 +132,24 @@ static const struct processing_id __maybe_unused processing_ids[] = {
+	SKS_PROCESSING_ID(SKS_CKM_SHA512_RSA_PKCS_PSS),
+	SKS_PROCESSING_ID(SKS_CKM_SHA224_RSA_PKCS),
+	SKS_PROCESSING_ID(SKS_CKM_SHA224_RSA_PKCS_PSS),
+	SKS_PROCESSING_ID(SKS_CKM_RSA_AES_KEY_WRAP),
```
this is not supported.

EDVTAZ replied:
I will remove it, thanks for pointing it out.

etienne-lms commented on core/lib/libtomcrypt/src/pk/rsa/rsa_verify_hash.c:
```diff
@@ -79,11 +79,12 @@ int rsa_verify_hash_ex(const unsigned char *sig,      unsigned long siglen,
   }
 
-  if (padding == LTC_PKCS_1_PSS) {
+  if (padding != LTC_PKCS_1_V1_5_NA1) {
```
As commented in **rsa_sign_hash.c**: previous sequence did not check hash for `LTC_PKCS_1_V1_5`. If this is an error, we should fix through a specific commit of propose straight to optee_os.

etienne-lms replied:
@EDVTAZ, do you want me to help on this and setup a patch series to propose this to mainline op-tee? This is about signing/verifing a pre-hashed image, right?

EDVTAZ replied:
I'm sorry for the delay, I was away last week. I can make the pull-request to mainline op-tee, the same way I made the pull requests to your fork, right?

To answer your question, yes it is about signing/verifying a pre-hashed image. The only problem was that there was no such operation in the GPD TEE API, that is why I added TEE_ALG_RSASSA_PKCS1_V1_5.

I will go through your other comments too now.

etienne-lms replied:
i think it would be nice pushing a P-R for this libutee and libtomcrypt change in the mainline OP-TEE.
Maintainer and I will help finding the best way to get this pre-hashed scheme supported.

EDVTAZ replied:
Ok, I will do that.

etienne-lms reviewed (commented):
Thanks for these.
It could be nice to have specific commits for specific changes, but this is not mandatory (time consuming). However, the libtommcrypt change should really be is a dedicated commit and propose straight the optee_os. (does "no ASN.1" really defines a prehashed pkcs1 v1.5  scheme?).

EDVTAZ committed:
Modifications according to discussion in PR

Signed-off-by: Gabor Szekely <szvgabor@gmail.com>

EDVTAZ commented:
Hi @etienne-lms ,
Thank you for your comments. If you agree with my comments about libtomcrypt, I can make a PR to mainline OP-TEE.

Regarding `LTC_PKCS1_V1_5_NA1`, I looked into it again, and it seems there was a bit of a misunderstanding on my part. None of the `TEE_AsymmetricSignDigest` operations actually do the hashing operation, they expect the data to be pre-hashed. However they need to know the hash function used because they include that information in the padding. However with `CKM_RSA_PKCS` we need `EMSA` padding (https://lists.randombit.net/pipermail/botan-devel/2008-November/000697.html), and `LTC_PKCS1_V1_5_NA1` does that.  
This means that currently all `CKM_*_RSA_PKCS` mechanisms in sks are missing a digest operation, because those mechanisms do hashing and signing too. I will be looking into that next.

And one last thing, I also added a PR for client side: https://github.com/etienne-lms/optee_client/pull/10. If you have time you could take a look at that as well. Thanks!

etienne-lms commented on ta_services/secure_key_services/include/sks_ta.h:
```diff
@@ -709,13 +709,15 @@ struct sks_attribute_head {
 /* Attribute specific values */
-#define SKS_UNDEFINED_ID			((uint32_t)0xFFFFFFFF)
+#define SKS_CK_UNAVAILABLE_INFORMATION		((uint32_t)0xFFFFFFFF)
+#define SKS_UNDEFINED_ID			SKS_CK_UNAVAILABLE_INFORMATION
```
thanks

etienne-lms commented on ta_services/secure_key_services/src/object.c:
```diff
@@ -521,11 +521,15 @@ uint32_t entry_find_objects_init(uintptr_t tee_session, TEE_Param *ctrl,
-			continue;
-		if (rv != SKS_OK)
-			goto bail;
+		/* If there are no attributes specified, we return
```
minor: let's use the op-tee coding style. multi-line comment go
```c
/*
 * <start comment from here>
 */
```

etienne-lms commented on ta_services/secure_key_services/src/object.c:
```diff
@@ -673,3 +677,141 @@ uint32_t entry_find_objects_final(uintptr_t tee_session, TEE_Param *ctrl,
+
+	obj = sks_handle2object(object_handle, session);
+	if (!obj)
+		return SKS_BAD_PARAM;
```
s/`return SKS_...`/`rv = SKS_...; goto bail;`/

etienne-lms commented on ta_services/secure_key_services/src/object.c:
```diff
@@ -673,3 +677,141 @@ uint32_t entry_find_objects_final(uintptr_t tee_session, TEE_Param *ctrl,
+	TEE_Free(template);
+	template = NULL;
+
+bail:
```
move `bail:` before `TEE_Free(template);`.

etienne-lms commented on ta_services/secure_key_services/src/object.c:
```diff
@@ -673,3 +677,141 @@ uint32_t entry_find_objects_final(uintptr_t tee_session, TEE_Param *ctrl,
+			attr_type_invalid = 1;
+		}
+		if (rv == SKS_SHORT_BUFFER)
+			buffer_too_small = 1;
```
I think sequence should be halted if `rc` is non of the above neither `SKS_OK`.

EDVTAZ replied:
Ok, I will correct this.

etienne-lms commented on ta_services/secure_key_services/src/pkcs11_attributes.c:
```diff
@@ -1485,3 +1485,50 @@ bool object_is_private(struct sks_attrs_head *head)
+	/* Generate 32 bit ID */
+	if (!set_id)
+		TEE_GenerateRandom(&id, sizeof(id));
+	else id = *set_id;
```
minor: split in 2 lines:
```c
else
	id = *set_id;
```

etienne-lms commented on ta_services/secure_key_services/src/pkcs11_attributes.c:
```diff
@@ -1485,3 +1485,50 @@ bool object_is_private(struct sks_attrs_head *head)
+	
+	/* Generate 32 bit ID */
+	if (!set_id)
+		TEE_GenerateRandom(&id, sizeof(id));
```
no issue with collision? should it insure ID unique?

EDVTAZ replied:
I added this, so OpenSSH would operate on the keys I generated (since it would refuse if there was no CKA_ID). It would be nice to check for collisions, but at the time I was writing this I couldn't come up with any easy way to do so and later forgot about it. The standard also doesn't seem to require us to ensure this:
> The CKA_ID attribute is intended as a means of distinguishing multiple public-key/private-key pairs held by the same subject (whether stored in the same token or not). (Since the keys are distinguished by subject name as well as identifier, it is possible that keys for different subjects may have the same CKA_ID value without introducing any ambiguity.)
> It is intended in the interests of interoperability that the subject name and key identifier for a certificate will be the same as those for the corresponding public and private keys (though it is not required that all be stored in the same token). However, Cryptoki does not enforce this association, or even the uniqueness of the key identifier for a given subject; in particular, an application may leave the key identifier empty.

In any case, I will look into how I can iterate through all objects present in the SKS and implement this check.

etienne-lms commented:
Nice changes. sorry for the late feedback.
Do you think you can rebuild this patch serie with specific topic changes?

EDVTAZ committed:
Changes according to PR discussion

Signed-off-by: Gabor Szekely <szvgabor@gmail.com>

EDVTAZ commented:
Thanks, I will rebuild it once I'm done with the cka_id collision detection, if that's all right with you. I will also make the P-R to mainline OPTEE. University started for me, so things may go a bit slower, but I'm working on it.

etienne-lms commented:
Thanks for your work. I think we can drop the CKA_ID collision issue. As your pointed at, collision are not forbidden and we cannot check against outer tokens :)
So a 32bit random is fine. To avoid collision, one may want 128bit UUID.

EDVTAZ committed:
Increase generated id to 128 bits

Since we are dropping collision detection here, I increased the CKA_ID
length, so accidental collision is lower probability.

etienne-lms commented:
merged through https://github.com/etienne-lms/optee_os/pull/12.
